### ISSUE TITLE
needle build --lora: exported .cact does not reproduce the adapter's behaviour in JAX (3.0.2, linux-arm64)

### ISSUE THREAD

**Issue by OriNachum** [cactus-compute/needle#134]
## Summary

A LoRA adapter trained with `needle finetune` gives the right tool calls when run in JAX, but the `.cact` archive that `needle build --lora` produces from it does not: through the engine it gets 2 of 30 of **its own training examples** right and answers almost every query with the same tool.

## Environment

- cactus-needle 3.0.2 from PyPI (`pip install 'cactus-needle[train]' 'jax[cuda13]'`), Python 3.12
- Linux aarch64 (NVIDIA DGX Spark, GB10), jax 0.11.2, `CudaDevice(id=0)`
- Engine: the `libneedle.so` the package fetches for this platform; base checkpoint `checkpoints/needle3.safetensors` as downloaded by `needle finetune` (HF snapshot `b274efcb211a9eef48c9a88da4b43bd569696a39`)

## Steps

1. Training file: 233 examples, 16 tools, each line `{"query": ..., "tools": [...16 schemas...], "answers": [{"name": ..., "arguments": {...}}]}` (about a third have `"answers": []`). Tool schemas are plain `{"name", "description", "parameters": {"type": "object", "properties": {...}}}`; 11 of the 16 tools take no arguments. Prompt length is 526 to 543 tokens and the longest target is 22, so nothing is cut by `--max-len 1024`.
2. `needle finetune train.jsonl --epochs 20 --lr 5e-4 --lora-rank 32 --lora-alpha 64 --seed 7 --val-split 0.1 --out adapter.safetensors`
   Validation loss falls from 2.93 to 0.157. The log says `numerics CQ W4 STE + A8 (matches export)`.
3. `needle build --lora adapter.safetensors --out tuned.cact` writes `63.47 MB  581 tensors  W4A8` (the published base archive is 35.3 MB).
4. Evaluate two ways on the same queries, greedy:
   - **JAX**: `load_checkpoint(base)`, `merge_lora(params, lora, adapter["scale"])`, `render_example(...)` for the prompt, `run.generate(..., temperature=0.0)`.
   - **Engine**: `Needle(weights="tuned.cact", tools=tools).complete(query)`.

## Result

| Evaluated on | JAX, merged LoRA | Engine, `tuned.cact` |
|---|---|---|
| Training examples (correct tool) | 6 checked: all 4 that expect a tool are correct | 2 of 30 |
| 80 held-back examples: 53 expecting a tool | 42 of 53 (stock in JAX: 11 of 53) | 6 of 80 over all 80 |

Typical engine output from the tuned archive, for queries whose training answer is `container_list` with no arguments:

```text
'containers?'                   -> service_status {"service": "docker"}
'Which containers are running?' -> service_status {"service": "container_list"}
'docker ps -a'                  -> service_status {"service": "ps"}
'List the containers'           -> service_status {"service": "container_list"}
```

The same merged weights in JAX return `[{"arguments":{},"name":"container_list"}]` for each of these.

## What was ruled out

- Prompt differences on the Python side: passing the tools as the same compact JSON string the trainer renders (`separators=(",", ":")`) and `auto_date=False` gives the identical 2 of 30 and 6 of 80.
- Truncation: every example fits in 1024 tokens (see step 1).
- The adapter not being applied: the archive's picks differ from the stock archive's on 29 of 80 queries, so the weights did change; they are just not the behaviour that was trained.

## Questions

- Is a W4A8 archive from `needle build --lora` expected to load correctly in the 3.0.x engine that ships the 2-bit base archive?
- Does the engine build its prompt the same way `finetune.render_example` does (no reasoning block, tools between the tools markers, then the query)? The stock model in JAX writes a reasoning block before its tool call, and the training examples here had none.

I can share the training file, the adapter and the two small evaluation scripts if that helps.

- nvsh (Claude)

**Comment by ParkiratS** [comment]
Looking into this

**Comment by OriNachum** [comment]
A correction to my numbers above, from a run that finished after I filed this.

The "2 of 30" and "6 of 80" figures in the table came from the two **prompt-matched** configurations (tools passed as the compact JSON string the trainer renders, and `auto_date=False`), not from the plain `Needle(weights="tuned.cact", tools=tools)` call the table's column header names. The same archive scored on a larger sample, all three ways:

| Engine, `tuned.cact` | Training examples right (of 60) | Held-back examples right (of 80) |
|---|---|---|
| `Needle(weights=..., tools=tools)` (defaults) | 10 | 12 |
| tools as the compact JSON string + `auto_date=False` | 3 | 7 |
| `auto_date=False` only | 3 | 7 |

So the default call does somewhat better than the prompt-matched ones, and both are far below the same merged weights in JAX (42 of 53 on the held-back examples that expect a tool). That the *matched* prompt does worse is itself a hint that the engine's prompt is not what `finetune.render_example` produces. The conclusion is unchanged: the exported archive does not reproduce the adapter's behaviour.

- nvsh (Claude)

**Comment by OriNachum** [comment]
@ParkiratS let me know if you want me to reproduce again, or guide me, or suggest a fix.

(I am unavailable for the next ~30h but can prioritize afterwards)

Kudos on the project and I'm happy using it in my project.

I'd love to add a citation on my Readme, too.

**Comment by HenryNdubuaku** [comment]
Thanks for the thorough report, this one was worth chasing. The engine is faithful to the archive; the gap is that your JAX baseline was full precision and the archive is 4-bit.

I trained a small adapter the same way and ran the same queries three ways:

| query | JAX float | JAX + W4 | engine .cact |
| --- | --- | --- | --- |
| containers? | container_list{} | container_list{} | container_list{} |
| Which containers are running? | container_list{"":{}} | container_list{} | container_list{} |
| docker ps -a | disk_usage{...} | service_status{"service":"docker ps -a"} | service_status{"service":"docker ps -a"} |
| show disk usage | disk_usage{} | service_status{"service":"disk_usage"} | service_status{"service":"disk_usage"} |
| is nginx up? | service_status{nginx} | service_status{nginx} | service_status{nginx} |

The engine matches JAX-at-W4 everywhere, and differs from JAX-float on exactly the queries quantisation changed, including your signature failure of the tool you wanted showing up as a `service_status` argument. I also checked the two quantisers against each other: training-time STE and the exporter agree to 0.002 per weight, so nothing is lost between them.

So the archive is doing what the numbers say; the float evaluation was just a more optimistic model than the one that ships. To make that visible, `needle finetune` now prints held-out accuracy scored on the same W4 weights the archive uses:

    epoch     12/12  loss 0.8047  val 0.8807
    accuracy  4/4 held-out calls exact, scored on the W4 weights the archive ships

If the quantised accuracy is lower than you need, more data and more epochs is the lever; 4 bits is the finest the .cact format carries today. Worth re-running your 233-example set with this to see the real number before export.
